src/satosa/satosa_config.py

```diff
@@ -44,26 +46,49 @@ def __init__(self, config):
 
         self._verify_dict(self._config)
 
+        self._load_plugins()
+        self._load_internal_attributes()
+
+    def _parse_config(self, config):
+        return next(
+            filter(
+                lambda conf: conf is not None,
+                map(
+                    lambda parser: parser(config),
+                    (self._load_dict, self._load_yaml),
+                ),
+            ),
+            None,
+        )
```

Comment:
I would write this, as:
```python
value = next(
    (
        conf
        for parser in [self._load_dict, self._load_yaml]
        for conf in [parser(config)]
        if conf
    ),
    None,
)
return value
```
it is cleaner, and more "pythonic"

Reply:
For "pythonic" (whatever this vague term means) I agree. For cleaner I think it's more subjective. I always found nested comprehensions to be less clear than nested maps (especially when nesting > 2)

Reply:
If I should put a debugger there I'm sure that there will be problems.
Pythonic is elegant, compact, winking and whatever but probably it could be better a simplified and debuggable approach. just put `import pdb; pdb.set_trace()` before that code before choosing, if we agree on a more debuggable and clear approach

Reply:
How would you put the debugger inside the comprehension to debug it? On the contrary I think it's easier to debug filter & map since each map call can be debugged on it's own, whereas a nested comprehension must be run fully or rewritten in smaller parts to be debugged.

Reply:
You simply can't! This is why it would be better to avoid that kind of generalization. This is just a personal view, I like the others code style, I also imitate these if needed and with curiosity. I'm jut getting older, that's all. Because of this I'd prefer a simple and very debuggable code. I've get into this thread to exchange some words about my personal view, now I've done, you can simply ignore me and go ahead :+1:

Reply:
Everyone's view is valuable. Feel free to share your thoughts @peppelinux 

---

I can perfectly debug this comprehension, exactly as I would debug the filter/map case. The reason you cannot have a debugger _inside_ a comprehension is a limitation of the language. And it is the exact same reason you cannot have a debugger _inside_ a lambda. Neither of the two forms can have a debugger _inside_ it.

If you really need to drop a debugger, it is very easy to convert the comprehension to a regular for-loop and place it under the right scope. While, you need to think more to convert the map/filter style into a regular for-loop.

The other way to place a debugger "in" there, is to create an actual function and have the debugger called inside the function - so, it is not in the comprehension or the map/filter, but in the function. Which again works the same for both styles.

---

Now, I don't think @peppelinux meant that map/filter is more debuggable; but that we should take debuggability into account before making generalizations. This is irrelevant to _where_ you can drop a debugger. What he says is that we should split complex actions, rather than try to squeeze them in these constructs; which is a valid point. (peppe, correct me if I'm wrong)

Part of debuggability is readability, and (in my eyes at least) comprehensions look and read better compared to map/filter. map/filter have their place and should be used, but they do not fit everywhere. From my experience with other developers, map/filter/lambdas are not regularly used and are not well understood, while even beginners can easily read comprehensions.

- Comprehensions read straight and well. map/filter/lambda result in more cognitive load: you have to think about the order of execution, you have to move in and out of the implicit loops, you have to parse lambdas.
- Comprehensions "nest" vertically. map/filter actually nest one inside the other.
- Comprehensions work with statements. map/filter need a callable to work with, which is why you need to create all those lambdas.

---

I used to use map/filter extensively, but in many cases readability is lost, everything feels complex, and people will prefer to not deal with this part of the code. This is not good. On the other hand, comprehensions do better for the big majority of cases.

There is only one thing that map/filter do better. And that is, they explicitly separate the application of the operation from the ordering of the items that the operation will be applied upon. Which in turn means, that a map/filter structure can be safely parallelized. This characteristic is important, but it is of little value here.

Reply:
I agree, you're right. During these years all we have used all the pythonics and whatever; when I code there come the time to choose a thing instead of another depending by the moment(!).Then when I come back to the code I always choose to refactor it following the simplest approach. I think that It would a good strategy merge in a separate branch a working code before refactor and merge It into master. Skoranda did something similar with ldap_store, this speed up the things. In programming languages we spent less time to code it than explain It, isn't so? My posts are the proof of this 🤭. Words are too often source of misunderstanding, where personal tastes can make ideologic conflicts, code is more neutral. If It works good then It could be refactored much better. These are complementaries steps, for all the other circumstances: more guru meditation 🖖

Reply:
> If you really need to drop a debugger, it is very easy to convert the comprehension to a regular for-loop and place it under the right scope. While, you need to think more to convert the map/filter style into a regular for-loop.
> 
You don't have to convert anything in the map/filter, you simply evaluate each part (the map part, or/and the filter part). There is no need to write any for loop.

Map/filter power is composition which cannot be achieved with comprehensions. Try having a 4 times nested comprehension (for complex constructs) to see how readable it is (I'm not sure even if it is achievable). Whereas map/filter does not change at all, you compose each map as an argument to the next map. List comprehensions have a limit since they are a syntactic construct whereas map is a higher order function that can be composed with much less limitation. 
As I said, I personally always found nested list comprehensions (especially crowded with if statements) more difficult to process (due to the change in order) and I've come to prefer map/filter exactly because I can compose them one part at a time and join them without having to think the order of the `for` statement like in comprehensions.

Readability for this specific discussion, depends on the experience one has with map/filter usage and concepts and it is a personal taste as well. If we had this conversation 10 years ago, mutating everything like crazy and having 10 levels deep for loops could be a normal thing that one would consider "readable". There is a reason concepts like map/filter and other functional concepts are being introduced and used more and more today even (if Guido wants to rant about how hard he finds map/filter/reduce and prefers comprehensions). The same thing can be said about the usage of `next`. One may be more familiar with a `for` loop with a `break` statement.

Reply:
> you simply evaluate each part

you cannot evaluate each part, as the two are nested. I would understand this if the code was in the form

```python
intermediate_result = map(foo, bar)
final_result = filter(baz, intermediate_result)
```

which directly translates to comprehensions

```python
intermediate_result = (foo(x) for x in bar)
final_result = (x for x in intermediate_result if baz(x))
```

> Map/filter power is composition

we agree on that.

> which cannot be achieved with comprehensions

I can't see why and I'm interested in understanding this.

> join them without having to think the order of the for statement

what is the order of the for statement?

---

For this specific snippet, there is nothing being composed. Creating lambdas on the fly is not composition. You would be composing if these were functions that were being reused to achieve a greater effect. Instead, this is just hardcoded behaviour in a map/filter construct. Just because someone is using map/filter/reduce, it does not mean that they're composing.

Reply:
> > you simply evaluate each part
> 
> you cannot evaluate each part, as the two are nested. I would understand this if the code was in the form

in order to debug the following
` (
        conf
        for parser in [self._load_dict, self._load_yaml]
        for conf in [parser(config)]
        if conf
 )`
you have to evaluate its parts. to evaluate them you have to write a `for` loop.

in order to evaluate this:

`
filter(lambda conf: conf is not None, map(lambda parser: parser(config), (self._load_dict, self._load_yaml)))
`
you evaluate `map(lambda parser: parser(config), (self._load_dict, self._load_yaml))` and you evaluate `filter(lambda conf: conf is not None` on the result. there is no need to think of any for loop, that's what I'm saying

> > Map/filter power is composition
> 
> we agree on that.
> 
> > which cannot be achieved with comprehensions
> 
> I can't see why and I'm interested in understanding this.
> 

I actually meant functional decomposition (splitting a problem into a set of functions). Composition is another feature that is not builtin in python (you have to create a function that will create composite function whereas in other languages there is a operator to do that)
Let me rephrase this. You cannot use them the way you use map/filter to process things. If you want to create list comprehensions with multiple checks and processing you have to write a new nested comprehension with extra variables.

For example:

```
def remainder_of_two(x):
    return int(x) % 2

def add_star(x):
    return "*{}".format(x)
```
I want to flatten a list of integers `a=[[1,2],[3,4]]`, transform them to string, get those that are odd and add a * character to them. I have the above functions.
With map/reduce/filter, I feed each operation into the other:

`list(map(add_star, filter(remainder_of_two, map(str, reduce(add, a)))))`
reading from right to left, I flatten the list, then I get the string representation of them, then I filter out those that are even, then I add the star character

With list comprehensions and if statements:
`[add_star(y) for y in [x for b in a for x in b if remainder_of_two(str(x)) != 0]]`

Now how do I read this to describe what is done? Can I read from right to left? No. Can I read from left to right? Probably, let's try: I add a star character to a variable that is the one that is contained in a list of other variables that the string representation of that sub-variable when applied the remainder_of_two function, returns != 0. Is this narrative correct? Not exactly because I never mentioned the flattening of the list of lists explicitly. That kind of processing with list comprehensions needs extra variables. map/filter/reduce are higher order functions and don't need those (unless you pass a lambda). (sidenote: until python 3 list comprehensions were leaking scope of their variables (that is, they were accessing outside variables instead of creating a local scope).



> > join them without having to think the order of the for statement
> 
> what is the order of the for statement?

As I said before, narrating the operation in one's head, feeding results in functions is simpler than having to keep variables in brain memory. In our case:
I want to apply a list of functions in a config, and if their result is truthish, I want to get the result:
` (
        conf
        for parser in [self._load_dict, self._load_yaml]
        for conf in [parser(config)]
        if conf
 )`
 How do we express this above? I have a result called conf that -skip to line 3- when applied to a variable called parser -skip to line 2- where parser is a loop over a list of functions, -skipt to line 4- if conf is truthish. What I mean by this extremely naive narration, is that in comprehensions you have to read the whole expression to understand what's happening. Whereas in map/filter in our case:
 `
 filter(lambda conf: conf is not None, map(lambda parser: parser(config), (self._load_dict, self._load_yaml)))
 `
 Reading for right to left: I have a list of functions, and I apply them to a config, and then the results I feed them to a check to see if they are truthish and keep those that are. 


> For this specific snippet, there is nothing being composed. Creating lambdas on the fly is not composition. You would be composing if these were functions that were being reused to achieve a greater effect. Instead, this is just hardcoded behaviour in a map/filter construct. Just because someone is using map/filter/reduce, it does not mean that they're composing.

Indeed there is no composition here since there aren't any composite functions

Reply:
> in order to debug the following [...] you have to evaluate its parts.

yes.

> to evaluate them you have to write a for loop.

no. As with map/filter you evaluate the different parts.

> in order to evaluate this: [...] 
> you evaluate map()
> and you evaluate filter() on the result

Which means that you evaluate its parts. 
And you can only do that if these are written separately, not nested.

I think that you are assuming that the way map/filter work in python, is the same as in most other languages, but this is not the case. The `map()` and `filter()` functions return `map` and `filter` objects respectively. `map` and `filter` objects are actually iterators that _yield_ (they work just like generators) items. This means that `map` and `filter` are not separately evaluated. When a function pulls an item (pull-functions are `next()`, `list()`, etc) the filter-fn will retrieve the next item from the given iterable; this iterable is the map-iterator; to get the next item of the map-iterator, the map-fn retrieves the next item from the given iterable; the map-fn applies its callable to the item, saves its state and returns a single result; then the filter-fn applies its callable to the item it got from the map-iterator, saves its state and returns a single result. The map/filter-fn never evaluate completely (or anything), if a pull-fn does not ask them to.

This works exactly as comprehensions work, and it is the reason you need to wrap the outermost function (filter in this case) with a pull-function, as you do with comprehensions.

In most other languages, map and filter do not return an iterator, but a complete result. That used to be the case for python2; but for python3 the effect is the same as with comprehensions.

---

> there is no need to think of any for loop, that's what I'm saying

yes. The reason I mentioned the transformation to the for loop, was to be able to invoke a debugger _inside_ the loop. 

If you're saying there is not reason to _look_ at the `for`-keyword of the comprehension, then you can hide it behind a function, just like map/filter do. Just because you do not see the for keyword, it does not mean that there is no loop. 

The only good thing with map/reduce is that you can replace them with versions that can parallelize the operation safely. (which is not needed here and in the majority of the cases)

---

> sidenote: until python 3 list comprehensions were leaking scope of their variables (that is, they were accessing outside variables instead of creating a local scope

this is completely irrelevant

---

> I want to 
> - flatten a list of integers a=[[1,2],[3,4]], 
> - transform them to string, 
> - get those that are odd 
> - and add a * character to them.

> With map/reduce/filter, I feed each operation into the other:
> ```
> list(
>     map(                       # add a * character
>         add_star,
>         filter(                # get odds
>             remainder_of_two, 
>             map(
>                 str,           # convert to str
>                 reduce(add, a) # flatten
>             )
>         )
>     )
> )
> ```

> reading from right to left ...

Reading from right to left is the opposite of what people do. To be precise, this reads inside-out (or bottom-top if you will).

> With list comprehensions and if statements:
> ```
> [
>     add_star(y)
>     for y in [
>         x
>         for b in a 
>         for x in b 
>         if remainder_of_two(
>             str(x)
>         )
>     ]
> ]
> ```

There is no need for this nesting. This should be written as

```
list(
    add_star(s)                        # add a * character
    for sublist in a for x in sublist  # flatten
    for s in [str(x)]                  # convert to str
    if remainder_of_two(s)             # check for odd
)
```

or even

```
list(
    v
    for sublist in a for x in sublist  # flatten
    for s in [str(x)]                  # convert to str
    if remainder_of_two(s)             # check for odd
    for v in [add_star(y)]             # add a * character
)
```

Reading is fine, from top to bottom.

---

> How do we express this above? I have a result called conf that -skip to line 3- ...

What you're doing here is the same as this:

> In order to read this:
> ```
> filter(lambda conf: conf is not None, map(lambda parser: parser(config), (self._load_dict, self._load_yaml)))
> ```
> you have to find the inner most function, then find the leftmost argument, then apply it to each item of the remaining arguments, then imagine the results, then skip to next outer function, then find the leftmost argument, then apply it to the remaining arguments including the one you imagined before, then imagine the results, and so on.

(which, again, is not true, because in practice map and filter return iterators.)

Likewise, I can read the comprehension as:

_for each parser, apply it to config and generate a conf, if that is truthish keep it_

I doubt that you find it easier to locate the inner-most function, and immediately reason about it, while you find it hard to read that comprehension.

In general, comprehensions read from top to bottom, left to right, and accumulate results in the very first line. Instead map/filter read inside-out, and demand understanding of the arguments.
I can read both, but I can read comprehensions easier, because the order of the statements is natural. Badly written code will read bad no matter if it is within a comprehension or a map/filter.

---

Having said all this, I do not think that there is _any other place_ where you would nest functions one inside the other like you do with map and filter. You would not do:
```
command3(command2(command1(arg1, arg2)))
```
instead you would split this into multiple lines that feed results into each other.
```
res1 = command1(arg1, arg2)
res2 = command2(res1)
res3 = command3(res2)
```
so, I cannot understand how this changes with map and filter, and how it is helpful to anyone to read it nested like that. Do people _really_ write nested code like that with map/filter/reduce? Do they try to convert their whole programs into a one-liner? Where does this come from?

> That kind of processing with list comprehensions needs extra variables

The only reason you do not have variables is because of this nesting. Variables is not a _bad thing_ that we should avoid. Instead, variables are pretty helpful to document in-code the intermediate results.

Reply:
> In most other languages, map and filter do not return an iterator, but a complete result. That used to be the case for python2; but for python3 the effect is the same as with comprehensions.

Well it depends. In Clojure for example maps are lazy (realized in chunks to be exact) whereas in Haskell they are not (they are non-strict which is not the same) although they act like the same

> 
> > there is no need to think of any for loop, that's what I'm saying
> 
> yes. The reason I mentioned the transformation to the for loop, was to be able to invoke a debugger _inside_ the loop.
> 
> If you're saying there is not reason to _look_ at the `for`-keyword of the comprehension, then you can hide it behind a function, just like map/filter do. Just because you do not see the for keyword, it does not mean that there is no loop.
> 
I said it to point the fact that you simply evaluate each part without rewriting it whereas in list comprehensions you have to rewrite it if it has many parts (i.e nested comprehension with if statement)

> > How do we express this above? I have a result called conf that -skip to line 3- ...
> 
> What you're doing here is the same as this:
> 
> > In order to read this:
> > ```
> > filter(lambda conf: conf is not None, map(lambda parser: parser(config), (self._load_dict, self._load_yaml)))
> > ```
> > 
> > 
> > you have to find the inner most function, then find the leftmost argument, then apply it to each item of the remaining arguments, then imagine the results, then skip to next outer function, then find the leftmost argument, then apply it to the remaining arguments including the one you imagined before, then imagine the results, and so on.
> 
> (which, again, is not true, because in practice map and filter return iterators.)
> 
> Likewise, I can read the comprehension as:
> 
> _for each parser, apply it to config and generate a conf, if that is truthish keep it_
> 
> I doubt that you find it easier to locate the inner-most function, and immediately reason about it, while you find it hard to read that comprehension.
> 
> In general, comprehensions read from top to bottom, left to right, and accumulate results in the very first line. Instead map/filter read inside-out, and demand understanding of the arguments.
> I can read both, but I can read comprehensions easier, because the order of the statements is natural. Badly written code will read bad no matter if it is within a comprehension or a map/filter.

Well from top to bottom in the example I read a v and the declaration for what `v` is, is at the bottom of the statement. That's what I meant that in comprehensions you use `goto` in your mind whereas in map/filter/reduce you read inside out and chain the parts together (you read, "I first do this, then I do that, then I do that etc" whereas in comprehensions you have to remember variable names and jump inside the comprehension to read). So comprehensions don't exactly read from top to bottom because if you read them from top to bottom you'll have to remember what `v` is. So you choose either `goto` and not top-to-bottom reading either remembering what that variable `v` in the first line will be defined 3-4 lines later.
> 
> Having said all this, I do not think that there is _any other place_ where you would nest functions one inside the other like you do with map and filter. You would not do:


> 
> so, I cannot understand how this changes with map and filter, and how it is helpful to anyone to read it nested like that. Do people _really_ write nested code like that with map/filter/reduce? Do they try to convert their whole programs into a one-liner? Where does this come from?
> 
They don't I'm simply mentioning this to underline why I think map/filter is more powerful; because it chains better

> > That kind of processing with list comprehensions needs extra variables
> 
> The only reason you do not have variables is because of this nesting. Variables is not a _bad thing_ that we should avoid. Instead, variables are pretty helpful to document in-code the intermediate results.

Variables is not a bad thing indeed. I simply mentioned them that you have to remember variables when reading a comprehension


I think we're beating a dead horse here, we will probably not reach an agreement. I think this ends up to personal taste. People using mostly Python might find comprehensions more intuitive because they are more exposed to them and the community & BDFL supports them. I started this discussion to underline the fact that "clearer" and "pythonic" are not objective terms. Especially "pythonic" can lead to assumptions that are counter-intuitive for other languages (i.e by default ordered dicts, assignment expressions as in pep572 etc)